Alerting: Add route matching functionality to package

gillesdemey opened the pull request:
**What is this feature?**

This PR adds the route matching algorithms to the alerting package so we can share it with plugins.

The new data structure will also record the matching journey so we can track and display the decision making process.

**Special notes for your reviewer:**

None

gillesdemey committed:
initial version

gillesdemey committed:
rename file

gillesdemey committed:
simplify matching results, pass tree details

gillesdemey committed:
update hook to return data structure that works with multiple trees

gillesdemey committed:
fixes to matching algorithm

gillesdemey committed:
update alerting pkg

gillesdemey committed:
update matching hook

gillesdemey committed:
update package data structure

gillesdemey committed:
minor updates

gillesdemey committed:
Merge remote-tracking branch 'origin' into alerting/pkg-add-route-matching-hook

gillesdemey committed:
track involved policies and include tests

gillesdemey committed:
record entire matching journey

gillesdemey committed:
refactor: update useMatchPolicies hook to convert RoutingTree to Route and adjust matching logic

gillesdemey committed:
test: refactor notification policy tests to use RouteFactory for route creation

gillesdemey committed:
export the routes fakes

gillesdemey committed:
updates to alerting pkg

gillesdemey committed:
refactor: enhance RoutingTree to Route conversion logic in notification policies

gillesdemey committed:
refactor: export USER_DEFINED_TREE_NAME and enhance utility imports in notification policies

gillesdemey committed:
lint and stuff

gillesdemey committed:
more linting sigh

gillesdemey committed:
add @grafana/data as dependency

gillesdemey committed:
Merge remote-tracking branch 'origin' into alerting/pkg-add-route-matching-hook

gillesdemey committed:
lockfile

konrad147 commented:
@cursor review

gillesdemey committed:
Merge remote-tracking branch 'origin/main' into alerting/pkg-add-route-matching-hook

gillesdemey committed:
Merge remote-tracking branch 'origin/main' into alerting/pkg-add-route-matching-hook

gillesdemey committed:
narrow type definition

gillesdemey committed:
Merge remote-tracking branch 'origin' into alerting/pkg-add-route-matching-hook

soniaAguilarPeiron reviewed (approved):
Added 2 non blocking comments, but the PR looks great! :shipit:

gillesdemey committed:
add additional tests to the package

gillesdemey committed:
update alerting pkg type

gillesdemey committed:
comments

gillesdemey commented on packages/grafana-alerting/src/grafana/matchers/utils.ts:
```diff
@@ -0,0 +1,127 @@
+
+  // for each matcher, check which label it matched for
+  matchers.forEach((matcher) => {
+    const matchingLabelIndex = labels.findIndex((label) => isLabelMatch(matcher, label));
```
Good question, I kept the behaviour from Grafana core for now.

Since we're working with a `LabelMatcher` with strongly defined set of string unions for the type we don't actually need this check – the type checker will force us to make sure the input matcher has a valid and supported operator type.

gillesdemey committed:
remove redundant throw

the type checker will force us to pass a matcher with a strict known
type operator

gillesdemey committed:
add test for invalid regex match

gillesdemey committed:
add test for "addUniqueIdentifier"

gillesdemey committed:
more test coverage

gillesdemey committed:
lint
